lib/backend/dynamo/dynamodbbk.go

```diff
@@ -383,15 +383,19 @@ func (b *Backend) GetRange(ctx context.Context, startKey []byte, endKey []byte,
 
 func (b *Backend) getAllRecords(ctx context.Context, startKey []byte, endKey []byte, limit int) (*getResult, error) {
 	var result getResult
+	limitRemaining := limit
 	// this code is being extra careful here not to introduce endless loop
 	// by some unfortunate series of events
 	for i := 0; i < backend.DefaultRangeLimit/100; i++ {
-		re, err := b.getRecords(ctx, prependPrefix(startKey), prependPrefix(endKey), limit, result.lastEvaluatedKey)
+		if limit > 0 {
+			limitRemaining = limit - len(result.records)
+		}
+		re, err := b.getRecords(ctx, prependPrefix(startKey), prependPrefix(endKey), limitRemaining, result.lastEvaluatedKey)
```

Comment:
Really really sorry for going back and forth on this. Now i am a little concerned how we are using the limit.

by [AWS documentation](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html), 
> [Limit](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_RequestSyntax)
The maximum number of items to evaluate (not necessarily the number of matching items).

Meaning the `Limit` is maximum number of items to be scanned, *pre-filtering*. The records we obtained is *after filtering*.


---
The only filtering we have in `getRecords` is `Expires >= :timestamp`. 

Let's use an example, say we have 10 records, and all first 9 records are expired.

If we use `Limit = 5` for this query, we wound **NOT** get any results back (since we scanned first 5, all expired). 

Even though, we have one matching record, we won't get it unless `Limit` is greater than 9. (or no limit)

---
So, in our case, is the `limit` parameter the **limit-of-items-to-scan-for-all-queries** or **limit-of-items-to-scan-per-query** or **limit-of-items-to-return**?

Reply:
The logic will work because we are relying on the `len(re.lastEvaluatedKey) == 0` check to verify if we scanned all elements thought settings the limit based on `limitRemaining = limit - len(result.records)` can introduce additional backend calls. For example if the limit was set to `1` and there are 1000 expired items the `getRecords` function will be invoked x1000 times. I have reverted this change.